projects/plugins/crm/includes/ZeroBSCRM.DAL3.php

```diff
@@ -3708,62 +3713,59 @@ public function getTagObjStats($args=array()){
-     *
-     * @return string tag slug
-     */
-    private function getGenericTagSlug($objTypeID=-1){
-
-        // if passed with obj type
-        if ($objTypeID > 0){
-
-            global $wpdb,$ZBSCRM_t;
-
-            // tag-*
-            $startingI = 1;
-
-            // try and retrieve last added (if any)
-            $potentialTag = $wpdb->get_var($wpdb->prepare("SELECT zbstag_slug FROM ".$ZBSCRM_t['tags']." WHERE zbstag_slug LIKE 'tag-%' AND zbstag_objtype = %d ORDER BY zbstag_slug DESC LIMIT 1",$objTypeID));
-
-            if (!empty($potentialTag)){
-
-                // try and retrieve $i
-                if (substr($potentialTag,0,4) == 'tag-'){
-
-                    $potentialI = (int)substr($potentialTag,4);
-                    if ($potentialI > 0) $startingI = $potentialI+1;
-
-                }
-
-            }
-
-            // now we theoretically will have 4 if there's a record 'tag-3'
-            // this is dangerously open to running giant loops, lets limit it to 1024 and field any feedback
-            // ... should only ever be called in the instance a tag slug can't be generated (chinese characters currently only case)
-            $i = $startingI;
-            while ($i <= 1024){
+	/**
+	 * Checks if a tag slug exists
+	 *
+	 * @param int    $obj_type_id Object type id.
+	 * @param string $slug Tag slug to check.
+	 *
+	 * @return string tag slug
+	 */
+	public function tag_slug_exists( int $obj_type_id, string $slug ) {
+		$slug_exists = $this->getTag(
+			-1,
+			array(
+				'objtype' => $obj_type_id,
+				'slug'    => $slug,
+				'onlyID'  => true,
+			)
+		);
+		return $slug_exists !== false;
+	}
 
-                // is this tag in use?
-                $existingTagID = (int)$this->getTag(-1,array(
-                                'objtype'   => $objTypeID,
-                                'slug'      => 'tag-'.$i,
-                                'onlyID'    => true
-                                ));
+	/**
+	 * Checks if a tag slug exists
+	 *
+	 * @param int    $obj_type_id Object type id.
+	 * @param string $slug Tag slug to check.
+	 *
+	 * @return string tag slug
+	 */
+	public function get_new_tag_slug( int $obj_type_id, string $slug ) {
+		global $wpdb, $ZBSCRM_t; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
+		$slug_exists = $this->tag_slug_exists( $obj_type_id, $slug );
 
-                if ($existingTagID <= 0) return 'tag-'.$i;
+		// slug as provided doesn't exist, so use that
+		if ( ! $slug_exists && $slug !== 'tag' ) {
+			return $slug;
+		}
 
-                $i++;
+		$slug_base = $slug . '-';
 
-            }
+		// get last iteration of tag slug
+		$sql_query = 'SELECT CAST(TRIM(LEADING %s FROM zbstag_slug) AS SIGNED) AS next_slug_id FROM ' . $ZBSCRM_t['tags'] . ' WHERE zbstag_slug LIKE CONCAT(%s,"%") AND zbstag_objtype = %d ORDER BY next_slug_id DESC LIMIT 1000'; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
```

Comment:
Why `DESC LIMIT 1000` and not `DESC LIMIT 1`?

Reply:
Because I left it in a debug state...or I wanted to make sure you were paying attention? 🤪 Addressed in 79ac83957e, and I also changed the column alias.